Write file sync

NewLunarFire opened the pull request:
I quickly implemented writeFileSync on top of import http since I had already written it. I did not implement permissions yet.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno?pullRequest=527) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno?pullRequest=527) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: NewLunarFire<br/>:white_check_mark: ry<br/>:white_check_mark: thgh<br/>:x: Tommy Savaria<br/><hr/>**Tommy Savaria** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno?pullRequest=527) it.</sub>

ry commented:
Please rebase on top of #530

ry commented:
Please add a test to `js/unit_tests.ts` ... just something small and simple to exercise the code - try writing to /blah.txt and check you get a permission denied.
We need tempdir functionality before we can really test it.

Permissions are set in `deno.flags.allowWrite` - if you want to add that functionality. If not, add a TODO please.

NewLunarFire commented:
By permisssions I meant the file mode, i.e. the field called perm in the fbs message, was not implemented. For flag permissions I would wait until #524 has landed.

NewLunarFire commented:
I cleaned up this branch. I rebased it off master and I took only the changes to implement writeFileSync. I used the existing test in js/unit_tests.ts except I create a randomly named directory instead of an actual tmpdir, which we will probably implement later.

ry commented on js/unit_tests.ts:
```diff
@@ -92,23 +92,27 @@ test(async function tests_readFileSync() {
-  writeFileSync("/tmp/test.txt", data, 0o666);
-  const dataRead = readFileSync("/tmp/test.txt");
+
+  // TODO @NewLunarFire: This test currently generates a random directory
```
Please use TODO(NewLunarFire)

ry commented on src/handlers.rs:
```diff
@@ -318,6 +324,18 @@ fn handle_read_file_sync(d: *const DenoC, filename: &str) {
   );
 }
 
+// Prototype https://github.com/denoland/deno/blob/golang/os.go#L171-L184
```
The link is incorrect - use: https://github.com/denoland/deno/blob/golang/os.go#L186-L195

ry commented on js/unit_tests.ts:
```diff
@@ -92,23 +92,27 @@ test(async function tests_readFileSync() {
+  // with creating an actual tmpdir in the future
+  const randomSuffix = Math.random().toString(36).substr(2);
+  const fn = "tmp-" + randomSuffix + "/test.txt";
+  writeFileSync(fn, data, 0o666);
```
I don't want to add a test right now that writes into the current directory. I'd rather you comment this test out and add a TODO for later.

What about a failing test?  Writing to "/baddir/badfile.txt" should fail.

NewLunarFire replied:
How can you make sure that the test will fail? If the user has the right to create a file in this directory, the test will pass.

ry replied:
`/baddir/` presumably doesn't exist... so it will fail

NewLunarFire commented:
Also to note that if creating the parent directory fails because the directory already exists, writeFileSync will fail completely.

NewLunarFire commented on src/fs.rs:
```diff
@@ -11,6 +11,12 @@ pub fn read_file_sync(path: &Path) -> std::io::Result<Vec<u8>> {
 }
 
+pub fn write_file_sync(path: &Path, content: &[u8]) -> std::io::Result<()> {
+  std::fs::create_dir_all(path.parent().unwrap()).and_then(|_| {
```
The reason writeFileSync created the directory before creating the file was to properly cache files in #530 . So we'll have to handle this elsewhere in #530 after this is merged.

ry replied:
I've already changed it in the http_imports branch (which I'm still working on). If you want to rebase this work on top of that branch, you're welcome to.
https://github.com/denoland/deno/blob/74b40792c357efb98ec0708f43f670583859b408/src/fs.rs#L14-L17

NewLunarFire replied:
You don't need to store f in a variable, much less a mutable one. You could write ```File::create(path).write_all(content)``` for the same result.

I won't rebase this on top of http imports, I think this should  be merged before http imports.
